warehouse/macaroons/caveats/__init__.py

```diff
@@ -104,7 +104,7 @@ class RequestUser(Caveat):
     user_id: StrictStr
 
     def verify(self, request: Request, context: Any, permission: str) -> Result:
-        if not isinstance(request.identity, User):
+        if not isinstance(request.identity, UserTokenContext):
             return Failure("token with user restriction without a user")
 
         if str(request.identity.id) != self.user_id:
```

Comment:
NB: I've added `UserTokenContext.id` so that this line continues to work. However, it could just as easily be this instead:

```suggestion
        if str(request.identity.user.id) != self.user_id:
```

...and that might be more idiomatic than giving the `UserTokenContext` a DB ID attribute.

Reply:
This looks like the only appearance of this pattern, so I'm going to remove the `UserTokenContext.id` property and go this route.